sql/catalyst/src/main/java/org/apache/spark/sql/connector/ExternalCommandRunnableProvider.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector;
+
+import org.apache.spark.annotation.Unstable;
+
+import java.util.Map;
+
+/**
+ * @since 3.0.0
+ */
+@Unstable
+public interface ExternalCommandRunnableProvider {
+  /**
+   * Execute a random DDL/DML command inside an external execution engine rather than Spark,
```

Comment:
Does this assume that the command runs on the driver side only, or a datasource can execute it in parallel on all available executors, and combine the results?

Reply:
Yeah, this only happens in driver. But, ideally, driver won't run the command itself, but delegates it to external execution engine depends on the datasource. For example, JDBC could establish a connection and run the command on the external DMBS.